### PR TITLE
RFC: nixos/release: Make aarch64-linux a supported system again

### PR DESCRIPTION
**Pull request by mweinelt**
The aarch64-linux builders on hydra have had a good track in the last year or so and I think it's a good idea to include them in the supported systems again.

This happens in preparation of improving the distributing of the installer ISOs for aarch64-linux systems and advertise them more prominently on the homepage.

###### Description of changes

This is an educated guess of what is needed to improve aarch64-linux support. Ideally we could leave out aarch64-specific channels starting with the 22.11 release cycle.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by samueldr — APPROVED** [review]
Given that the only reason it's not is that [the previous hydra evaluator had trouble with it enabled](https://github.com/NixOS/nixpkgs/commit/1bfe8f189bd25de73340635d4e94673000f273ec), I think it should be done.

The evaluator was re-written since.

**Comment by vcunat** [comment]
I think we should unify this with stable releases.  There we have limited support in the main channel and an additional aarch64-linux channel.  I'm not saying it needs to be that way, but I can't see why handle stable and unstable differently.

**Comment by samueldr** [comment]
I'm not even sure why there are different jobsets. It causes issues when the updates are out of sync and you want to update the channel for a mixed deployment.

**Comment by vcunat** [comment]
22.05 will end relatively soon, so I think we can keep that one as it is.  But for future I believe it would make sense to unify the approaches in unstable and stable (stable e.g. since 22.11, if we decide soon).

As the build farm is now, with on-demand scaling of builders for both `*-linux`, I think it should be fine to include both platforms in the main big channel.  (and hope there won't be difficult regressions specific to aarch64-linux)  But if we go that way, I see no sense in keeping the `nixos-22.11-aarch64` jobset and channel, except perhaps as redirect.  I mean, if we wanted a separate channel in addition to one with both platforms, it would surely be for x86_64, right? (that would open the question of which should be the default, too)

**Comment by samueldr** [comment]
> I mean, if we wanted a separate channel in addition to one with both platforms, it would surely be for x86_64, right? (that would open the question of which should be the default, too)

In my opinion, if it was to happen, we'd want the "unqualified" channel to wait for all systems, and as many qualified channel names as we have supported systems. Weird in-betweens with only one arch exposed this way might be confusing for end-users.

Though I don't think there is any need for per-system channels, but I also don't have any data proving it either way.

**Comment by vcunat** [comment]
I wouldn't be keen on creating so many jobs(ets), as the central Hydra element isn't very efficient there.  Caching a build from another jobset apparently needs nontrivial work, e.g. after `staging-next` merge it can take several hours to just do that for those roughly 240k jobs we have in the jobsets attached to the `master` branch.

**Comment by samueldr** [comment]
I'm not either. We probably don't need them.

I'm mostly saying that if we ended-up needing them, we'd need to have one for each supported arch, no bizarro in-between.

**Comment by mweinelt** [comment]
Thinking of the added eval cost alone makes me dizzy.

**Comment by vcunat** [comment]
`-small` won't work as suggested in this PR, because some of the channel-critical tests won't evaluate:
```
in job ‘nixos.tests.installer.separateBoot.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.lvm.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.simple.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64
```

**Comment by vcunat** [comment]
Maybe we should test this branch on Hydra job(s) before merging, so that we don't immediately block the channel(s).

**Comment by mweinelt** [comment]
Moved the mentioned tests over to x86_64-linux only for -small. Testing the jobsets before prime-time sgtm.

File: nixos/release-small.nix
Comment (vcunat) [inline nixos/release-small.nix:89]: No `limitedSupportedSystems` in this file so far.

Comment (vcunat) [reply nixos/release-small.nix:89]: So just this line?
```nix
        (pkgs.lib.intersectLists systems supportedSystems);
```

Comment (mweinelt) [reply nixos/release-small.nix]: Hah, yeah. I was in a hurry and didn't test this properly. Fix pushed.

**Comment by mweinelt** [comment]
> Maybe we should test this branch on Hydra job(s) before merging, so that we don't immediately block the channel(s).

@vcunat What's the hold up here?

**Comment by vcunat** [comment]
Oh, I didn't understand that you want it already.  Let's start with small:
https://hydra.nixos.org/jobset/nixos/pr-193600-aarch64-support#tabs-errors

**Comment by mweinelt** [comment]
My hope is that we can test drive this early enough so we can get comfortable with it on unstable and then make an informed decision around channel creation time.

File: nixos/release-small.nix
Comment (misuzu) [inline nixos/release-small.nix:109]: Maybe we should also include [`nixos.tests.boot.uefiCdrom`](https://hydra.nixos.org/build/193876884).

Comment (mweinelt) [reply nixos/release-small.nix:109]: Sounds good to me.

**Comment by mweinelt** [comment]
These tests are currently unconditionally inherited, but not part of the tested set, so they log an eval error, but they don't make the evaluation fail.

```
in job ‘nixos.tests.installer.lvm.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.separateBoot.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.simple.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64
```

**Comment by vcunat** [comment]
I assume that the jobset is just for evaluation checks, so I cancelled the builds.

**Comment by mweinelt** [comment]
> `-small` won't work as suggested in this PR, because some of the channel-critical tests won't evaluate:
> 
> ```
> in job ‘nixos.tests.installer.separateBoot.aarch64-linux’:
> error: Non-EFI boot methods are only supported on i686 / x86_64
> 
> in job ‘nixos.tests.installer.lvm.aarch64-linux’:
> error: Non-EFI boot methods are only supported on i686 / x86_64
> 
> in job ‘nixos.tests.installer.simple.aarch64-linux’:
> error: Non-EFI boot methods are only supported on i686 / x86_64
> ```

They are not part of the tested set, so primarily noise. I'm not sure how I could match the system in this attribute set:

https://github.com/NixOS/nixpkgs/blob/a7582289f5df0ae48af4b6d291da2c0684b73d2c/nixos/release-small.nix#L48-L53

I think it could be ok to leave this as is.

> I assume that the jobset is just for evaluation checks, so I cancelled the builds.

Yeah, eval seems to have worked just fine. Is this good enough as is?

**Comment by vcunat** [comment]
I can't see anything blocking this.

**Review by vcunat — APPROVED** [review]
I can't see anything blocking this PR.

**Review by samueldr — APPROVED** [review]

**Comment by vcunat** [comment]
I realized a bit late, but... I think the worst about this PR are concerns about evaluation performance on Hydra.  (I don't consider that a deal-breaker, though.)  The evals including NixOS tests are expensive and while we've apparently solved memory issues, eval now took more than an hour: https://hydra.nixos.org/eval/1783395

**Comment by vcunat** [comment]
Actually, on a [quick glance](https://monitoring.nixos.org/grafana/d/hkRCcV0mk/instance-metrics?orgId=1&refresh=30s&from=now-24h&to=now) during performing the evaluation, the `rhea` machine had a few peaks of RAM usage above 95%, so I hope we won't be running into this again.  You know, packages and tests get added continuously, so that by itself makes the requirements grow over time.

**Pull request merged**
